shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java

```diff
@@ -781,17 +784,17 @@ void onUserLeaveHint() {
   void onTrimMemory(int level) {
     ensureAlive();
     if (flutterEngine != null) {
-      // This is always an indication that the Dart VM should collect memory
-      // and free any unneeded resources.
-      flutterEngine.getDartExecutor().notifyLowMemoryWarning();
       // Use a trim level delivered while the application is running so the
       // framework has a chance to react to the notification.
-      if (level == TRIM_MEMORY_RUNNING_LOW) {
-        Log.v(TAG, "Forwarding onTrimMemory() to FlutterEngine. Level: " + level);
+      // Avoid being too aggressive before the first frame is rendered. If it is
+      // not at least running critical, we should avoid delaying the frame for
+      // an overly aggressive GC.
+      boolean trim =
+          isFirstFrameRendered ? level >= TRIM_MEMORY_RUNNING_LOW : level > TRIM_MEMORY_RUNNING_LOW;
+      if (trim) {
+        flutterEngine.getDartExecutor().notifyLowMemoryWarning();
         flutterEngine.getSystemChannel().sendMemoryPressureWarning();
```

Comment:
so we're running it more now? From being only in TRIM_MEMORY_RUNNING_LOW to all levels above TRIM_MEMORY_RUNNING_LOW?

Reply:
Correct.

I'm not sure why the previous behavior was the way it was, but in this case we'll now run it when the app goes to background. I can move this to a separate change if you think that's worthwhile.

What's bad right now is that I suspect you can get `RUNNING_CRITICAL` and never hit `RUNNING_LOW`, or you might go to background and get one of the background related ones without ever hitting a `RUNNING_*` one.

Reply:
Sorry, to clarify:

We're running `flutterEngine.getDartExecutor().notifyLowMemoryWarning()` less now. This is the bigger culprit since it almost always will force an expensive GC.

We're running `flutterEngine.getSystemChannel.sendMemoryPressureWarning()` more now, which is less serious because it does not by itself trigger GC, just triggers a platform message to the Dart application to do cleanup.

Reply:
right. I think one PR is ok. Just to check on potential expectations, is that with this PR, we might increase jank (more GC events going into the framework), and we might decrease OOM (since we're not missing some signals now). Is that right?

Reply:
That sounds right to me